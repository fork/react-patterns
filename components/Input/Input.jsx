import PropTypes from 'prop-types';
import React from 'react';

import StyledInput from './Input.style';

const Input = ({
  autocomplete,
  disabled,
  error,
  errorMessage,
  hideLabel,
  id,
  label,
  onChange,
  pattern,
  placeholder,
  required,
  success,
  type,
  value,
  ...props
}) => (
  <StyledInput disabled={disabled} error={error} success={success} hideLabel={hideLabel} {...props}>
    <label htmlFor={id}>
      <span className="input__labelText">{label}</span>
      {required && <span className="input__required">*</span>}
    </label>

    <div className="input__wrapper">
      <input
        id={id}
        name={id}
        type={type}
        placeholder={placeholder}
        onChange={onChange}
        aria-invalid={error}
        aria-required={required}
        disabled={disabled}
        pattern={pattern}
        autoComplete={autocomplete}
        value={value}
      />
    </div>
    {error && errorMessage && <span className="input__error">{errorMessage}</span>}
  </StyledInput>
);

Input.propTypes = {
  autocomplete: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  hideLabel: PropTypes.bool,
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  /**
   Regex pattern for validation
  */
  pattern: PropTypes.string,
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  success: PropTypes.bool,
  type: PropTypes.string,
  value: PropTypes.string
};

Input.defaultProps = {
  autocomplete: 'off',
  disabled: false,
  error: false,
  errorMessage: '',
  hideLabel: false,
  onChange: null,
  pattern: null,
  placeholder: '',
  required: false,
  success: false,
  type: 'text',
  value: ''
};

export default Input;
