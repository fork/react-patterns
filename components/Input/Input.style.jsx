import styled, { css } from 'styled-components';
import { color, space, hideVisually } from '../../stylesheets';

const StyledInput = styled.div`
  label {
    display: block;
    font-weight: bold;
    color: ${color('black')};
    line-height: 1;
    margin-bottom: ${space('s')};
    ${props => props.hideLabel && hideVisually()}
  }

  input {
    width: 100%;
    outline: none;
    background: transparent;
    color: ${color('black')};
    font-size: inherit;
    font-weight: normal;
    border: 1px solid ${color('black')};
    padding: ${space('s')};

    &::placeholder {
      font-weight: normal;
    }

    &::-ms-input-placeholder {
      font-weight: normal;
    }

    &:-ms-input-placeholder {
      font-weight: normal;
    }

    &[disabled] {
      opacity: 0.5;
    }

    &:focus {
      border-color: ${color('secondary')};
    }
  }

  .input__wrapper {
    position: relative;
  }

  .input__error {
    display: block;
    color: ${color('highlight')};
  }

  ${props =>
    props.success &&
    css`
      label {
        color: ${color('secondary')};
      }

      input {
        border-color: ${color('secondary')};
      }
    `}

  ${props =>
    props.error &&
    css`
      label {
        color: ${color('highlight')};
      }

      input {
        border-color: ${color('highlight')};
      }
    `}
`;

export default StyledInput;
