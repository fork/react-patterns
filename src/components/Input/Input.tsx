import React from 'react';

import StyledInput from './Input.style';

export type InputProps = {
  autocomplete?: string;
  disabled?: boolean;
  error?: boolean;
  errorMessage?: string;
  hideLabel?: boolean;
  id: string;
  label: string;
  onBlur?: (ev: React.FocusEvent<HTMLInputElement>) => void;
  onChange?: (ev: React.ChangeEvent<HTMLInputElement>) => void;
  onFocus?: (ev: React.FocusEvent<HTMLInputElement>) => void;
  /**
   Regex pattern for validation
  */
  pattern?: string;
  placeholder?: string;
  required?: boolean;
  success?: boolean;
  type?: string;
  value?: string;
};

const Input: React.FC<InputProps> = ({
  autocomplete,
  disabled,
  error,
  errorMessage,
  hideLabel,
  id,
  label,
  onBlur,
  onChange,
  onFocus,
  pattern,
  placeholder,
  required,
  success,
  type = 'text',
  value
}: InputProps) => (
  <StyledInput error={error} success={success} hideLabel={hideLabel}>
    <label htmlFor={id}>
      <span className="input__labelText">{label}</span>
      {required && <span className="input__required">*</span>}
    </label>

    <div className="input__wrapper">
      <input
        id={id}
        name={id}
        type={type}
        placeholder={placeholder}
        onChange={onChange}
        onFocus={onFocus}
        onBlur={onBlur}
        aria-invalid={error}
        aria-required={required}
        disabled={disabled}
        pattern={pattern}
        autoComplete={autocomplete}
        value={value}
      />
    </div>
    {error && errorMessage && <span className="input__error">{errorMessage}</span>}
  </StyledInput>
);

export default Input;
