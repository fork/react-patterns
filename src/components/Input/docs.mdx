import { Story, Canvas } from '@storybook/addon-docs';

# Input

Input-Felder erlauben den Nutzer:innen die Eingabe von Text. Sie tauchen in der Regel in Formularen oder Dialogen auf.

<Canvas>
  <Story id="components-input-input--without-value" />
</Canvas>

## Anatomie

Ein Input besteht aus dem eigentlichen Eingabefeld, einem Label und ein Platzhalter-Text (optional).

<!-- Optional: schematische Darstellung -->

1. **Label**: Beschreibt den einzugebenen Inhalt
2. **Input**: Eingabefeld für den Text des:der Nutzer:in
3. **Platzhalter** (optional): Kann verwendet werden um ein Beispielinhalt zu kommunizieren.

<!--
## Platzierung

Hier kann beschrieben werden, wie ein Input in verschiedenen Kontexten platziert wird.
-->

<!--
## Inhalt

Hier können Regeln für die Inhalte von Inputs definiert werden, z.B. Wording von Labels.
-->
