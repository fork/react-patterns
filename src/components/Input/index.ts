export type { InputProps } from './Input';
export { default } from './Input';
