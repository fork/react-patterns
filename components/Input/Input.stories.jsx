import React, { useState } from 'react';

import Input from './Input';

import props from './data/00-default';

export default {
  title: 'Input',
  component: Input
};

export const withoutValue = () => <Input {...props} />;
export const withValue = () => <Input {...props} value="Hans Peter" />;
export const withoutLabel = () => <Input {...props} hideLabel />;
export const error = () => <Input {...props} value="Hans Peter" error />;
export const success = () => <Input {...props} value="Hans Peter" success />;
export const disabled = () => <Input {...props} disabled />;
export const withState = () => {
  const [value, setValue] = useState('');
  return <Input {...props} value={value} onChange={e => setValue(e.target.value)} />;
};
