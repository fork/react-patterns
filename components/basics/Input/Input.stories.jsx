import React, { useState } from 'react';

import Input from './Input';

import props from './data/00-default';

export default {
  title: 'Basics/Input',
  component: Input
};

export const withoutValue = () => <Input {...props} />;
export const withValue = () => <Input {...props} value="Johannes" />;
export const withoutLabel = () => <Input {...props} hideLabel />;
export const error = () => <Input {...props} value="Johannes" error />;
export const success = () => <Input {...props} value="Johannes" success />;
export const disabled = () => <Input {...props} disabled />;
export const withState = () => {
  const [value, setValue] = useState('');
  return <Input {...props} value={value} onChange={e => setValue(e.target.value)} />;
};
