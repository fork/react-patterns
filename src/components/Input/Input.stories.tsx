import React, { useState } from 'react';

import Input, { InputProps } from './Input';
import docs from './docs.mdx';

export default {
  title: 'Components/Input/Input',
  component: Input,
  parameters: {
    docs: {
      page: docs
    }
  }
};

const props: InputProps = {
  id: 'firstName',
  label: 'Vorname',
  value: '',
  required: true,
  type: 'text',
  placeholder: 'Gebe einen Vornamen ein...',
  onChange: () => console.log('Change'),
  onFocus: () => console.log('Focus'),
  onBlur: () => console.log('Blur'),
  error: false,
  success: false,
  pattern: null,
  disabled: false,
  autocomplete: null,
  errorMessage: 'Gebe einen richtigen Vornamen ein.'
};

export const withoutValue = () => <Input {...props} />;
export const withValue = () => <Input {...props} value="Hans Peter" />;
export const withoutLabel = () => <Input {...props} hideLabel />;
export const error = () => <Input {...props} value="Hans Peter" error />;
export const success = () => <Input {...props} value="Hans Peter" success />;
export const disabled = () => <Input {...props} disabled />;
export const withList = () => {
  const [value, setValue] = useState('');
  return (
    <Input
      {...props}
      value={value}
      onChange={e => setValue(e.target.value)}
      list={{
        id: 'names',
        options: [
          'Andreas',
          'Holger',
          'Jan',
          'Joachim',
          'Johannes',
          'Jonas',
          'Julia',
          'Leonard',
          'Martin',
          'Olaf',
          'Sascha',
          'Siegfried',
          'Wolf'
        ]
      }}
    />
  );
};
export const withState = () => {
  const [value, setValue] = useState('');
  return <Input {...props} value={value} onChange={e => setValue(e.target.value)} />;
};
