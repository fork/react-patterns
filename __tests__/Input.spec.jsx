import React from 'react';
import { shallow } from 'enzyme';

import Input from '../components/Input';

describe('Input', () => {
  test('Input should render correctly', () => {
    const component = shallow(<Input label="Input Label" id="1" />);

    expect(component).toMatchSnapshot();
  });

  test('Input with hidden label should render correctly', () => {
    const component = shallow(<Input hideLabel label="Input Label" id="1" />);

    expect(component).toMatchSnapshot();
  });

  test('Input with default value should render correctly', () => {
    const component = shallow(<Input hideLabel label="Input Label" value="true" id="1" />);

    expect(component).toMatchSnapshot();
  });
});
