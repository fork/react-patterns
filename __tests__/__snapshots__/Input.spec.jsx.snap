// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Input should render correctly 1`] = `
<styled.div
  disabled={false}
  error={false}
  hideLabel={false}
  success={false}
>
  <label
    htmlFor="1"
  >
    <span
      className="input__labelText"
    >
      Input Label
    </span>
  </label>
  <div
    className="input__wrapper"
  >
    <input
      aria-invalid={false}
      aria-required={false}
      autoComplete="off"
      disabled={false}
      id="1"
      name="1"
      onChange={null}
      pattern={null}
      placeholder=""
      type="text"
      value=""
    />
  </div>
</styled.div>
`;

exports[`Input Input with default value should render correctly 1`] = `
<styled.div
  disabled={false}
  error={false}
  hideLabel={true}
  success={false}
>
  <label
    htmlFor="1"
  >
    <span
      className="input__labelText"
    >
      Input Label
    </span>
  </label>
  <div
    className="input__wrapper"
  >
    <input
      aria-invalid={false}
      aria-required={false}
      autoComplete="off"
      disabled={false}
      id="1"
      name="1"
      onChange={null}
      pattern={null}
      placeholder=""
      type="text"
      value="true"
    />
  </div>
</styled.div>
`;

exports[`Input Input with hidden label should render correctly 1`] = `
<styled.div
  disabled={false}
  error={false}
  hideLabel={true}
  success={false}
>
  <label
    htmlFor="1"
  >
    <span
      className="input__labelText"
    >
      Input Label
    </span>
  </label>
  <div
    className="input__wrapper"
  >
    <input
      aria-invalid={false}
      aria-required={false}
      autoComplete="off"
      disabled={false}
      id="1"
      name="1"
      onChange={null}
      pattern={null}
      placeholder=""
      type="text"
      value=""
    />
  </div>
</styled.div>
`;
